File: client/src/modules/models/Order.js
/**
 * Created by CPU60133_LOCAL on 11/29/2017.
 */

var Order = cc.Class.extend({

    orderId: 0,
    itemList: [],
    orderPrice: 0,
    orderExp: 0,
    waittingTime: null,

    ctor: function (orderId, itemList, orderPrice, orderExp) {
        //
        //this._super();

        this.orderId = orderId;

        this.createOrder(itemList, orderPrice, orderExp);
    },


    createOrder: function (itemList, orderPrice, orderExp) {
        this.itemList = itemList;
        this.orderPrice = orderPrice;
        this.orderExp = orderExp;

        this.waittingTime = null;
    },

    getOrderId: function () {
        return this.orderId;
    },
    getItemList: function () {
        //StorageItem
        return this.itemList;
    },

    //
    getFinishWaittingTime: function () {
        if (this.itemList == []){
            return null;
        }
        if (this.waittingTime == null){
            return null;
        }

        var parseTime = this.waittingTime.getTime();
        var finishTime = new Date();
        finishTime.setTime(parseTime + this.getRemainTime(user.getLevel()) * 60 * 1000);
        //finishTime.setTime(parseTime + 6000);     //HERE IS TEST

        return finishTime;
    },

    getRemainTime: function (level) {
        if (level < 11){
            return 3;
        } else if (level < 16){
            return 5;
        } else if (level < 21){
            return 10;
        } else if (level < 27){
            return 15;
        } else {
            return 30;
        }
    },

    checkStatus: function () {
        if (this.itemList.length > 0){
            return OrderStatusTypes.REALIZABLE;
        } else {
            return OrderStatusTypes.WAITTING;
        }
    },

    //
    checkCondition: function () {
        if (this.checkStatus() == OrderStatusTypes.WAITTING){
            return false;
        }

        var missingItem = [];
        for (var i = 0; i < this.itemList.length; i++){
            if (this.itemList[i].quantity > user.getAsset().getQuantityOfTwoStorageByProductId(this.itemList[i].typeItem)){
                missingItem.push(new StorageItem(this.itemList[i].typeItem,
                    this.itemList[i].quantity - user.getAsset().getQuantityOfTwoStorageByProductId(this.itemList[i].typeItem)));

            }
        }
        // return true;
        return (missingItem.length == 0) ? true : missingItem;
    },

    checkRemainItem: function () {
        var remainItem = [];
        for (var i = 0; i < this.itemList.length; i++){
            if (this.itemList[i].quantity <= user.getAsset().getQuantityOfTwoStorageByProductId(this.itemList[i].typeItem)){
                remainItem.push(new StorageItem(this.itemList[i].typeItem, this.itemList[i].quantity));
            } else if (user.getAsset().getQuantityOfTwoStorageByProductId(this.itemList[i].typeItem) > 0){
                remainItem.push(new StorageItem(this.itemList[i].typeItem, user.getAsset().getQuantityOfTwoStorageByProductId(this.itemList[i].typeItem)));
            }
        }

        return remainItem;
    },

//
    makeOrder: function () {
        if (this.checkCondition() == true){
            for (var i = 0; i < this.itemList.length; i++){
                user.getAsset().getFoodStorage().takeItem(this.itemList[i].typeItem, this.itemList[i].quantity);
                user.getAsset().getWarehouse().takeItem(this.itemList[i].typeItem, this.itemList[i].quantity);
            }

            //user.addGold(this.orderPrice);
            //user.addExp(this.orderExp);
            user.getAsset().getCar().updateDelivery(this.orderPrice, this.orderExp);

            return true;
        }
        return false;
    },

    makeOrderByRuby: function (rubyBuy) {

        if (!user.reduceRuby(rubyBuy)){
            return false;
        }

        var remainItem = this.checkRemainItem();
        for (var i = 0; i < remainItem.length; i++){
            user.getAsset().getFoodStorage().takeItem(remainItem[i].typeItem, remainItem[i].quantity);
            user.getAsset().getWarehouse().takeItem(remainItem[i].typeItem, remainItem[i].quantity);
        }

        //user.addGold(this.orderPrice);
        //user.addExp(this.orderExp);
        user.getAsset().getCar().updateDelivery(this.orderPrice, this.orderExp);

        return true;

    },


    cancelOrder: function () {
        //
        if (this.checkStatus() == OrderStatusTypes.REALIZABLE){
            this.itemList = [];
            this.orderPrice = 0;
            this.orderExp = 0;

            this.waittingTime = new Date();

            //
            return true;
        }
        return false;
    },

    boostWait: function () {
        if (user.reduceRuby(3)){
            return true;
        }
        return false;
    }


});

File: server/src/model/Order.java
package model;

import config.enums.ErrorLog;

import config.jsonobject.ProductConfig;

import config.utils.OrderUtil;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import util.database.DataModel;

public class Order extends DataModel{
    
    private int orderId;
    private List<StorageItem> itemList;
    
    private int orderPrice;
    private int orderExp;
    
    private long waittingTime;
    
    
    public Order(){
        super();
        
    }
    
    public Order(ZPUserInfo user, int level) {
        super();
        
        this.waittingTime = 0;
        this.createOrder(user, level);
    }
    


    public short createOrder(ZPUserInfo user, int level){
        if ((this.waittingTime + OrderUtil.getRemainTime(level) * 60 * 1000 - 5000) <= new Date().getTime()){
            this.setItemList(user, level);
            this.setOrderPrice(level);
            this.setOrderExp(level);
            
            this.waittingTime = 0;
            
            return ErrorLog.SUCCESS.getValue();
        }
        return ErrorLog.SUCCESS.getValue();
    }
    

    public void setOrderId(int orderId) {
        this.orderId = orderId;
    }

    public int getOrderId() {
        return orderId;
    }

//
    public void setItemList(ZPUserInfo user, int level) {
        List<ProductConfig> productList = OrderUtil.randomTypeProduct(user, level);
        this.itemList = OrderUtil.randomQuantityOfProductList(level, productList);
        
    }

    public List<StorageItem> getItemList() {
        return itemList;
    }


//
    public void setOrderPrice(int level) {
        this.orderPrice = OrderUtil.getOrderPrice(level, itemList);
        
    }

    public int getOrderPrice() {
        return orderPrice;
    }

    public void setOrderExp(int level) {
        this.orderExp = OrderUtil.getOrderExp(level, this.itemList);
        
    }

    public int getOrderExp() {
        return orderExp;
    }

    public void setWaittingTime(long waittingTime) {
        this.waittingTime = waittingTime;
    }

    public long getWaittingTime() {
        return waittingTime;
    }
    
    
    
    //
    public List<StorageItem> checkMissingItem(ZPUserInfo user){
        
        List<StorageItem> missingItemList = new ArrayList<>();
        for (int i = 0; i < this.itemList.size(); i++){
            if (this.itemList.get(i).getQuantity() > user.getAsset().getQuantityOfTwoStorageByProductId(this.itemList.get(i).getTypeItem())){
                missingItemList.add(new StorageItem(this.itemList.get(i).getTypeItem(),
                    this.itemList.get(i).getQuantity() - user.getAsset().getQuantityOfTwoStorageByProductId(this.itemList.get(i).getTypeItem())));
            }
        }
        return missingItemList;
    }
    
    public List<StorageItem> checkRemainItem(ZPUserInfo user){
        
        List<StorageItem> remainItemList = new ArrayList<>();
        for (int i = 0; i < this.itemList.size(); i++){
            if (this.itemList.get(i).getQuantity() <= user.getAsset().getQuantityOfTwoStorageByProductId(this.itemList.get(i).getTypeItem())){
                remainItemList.add(new StorageItem(this.itemList.get(i).getTypeItem(), this.itemList.get(i).getQuantity()));
            } else if (user.getAsset().getQuantityOfTwoStorageByProductId(this.itemList.get(i).getTypeItem()) > 0){
                remainItemList.add(new StorageItem(this.itemList.get(i).getTypeItem(), user.getAsset().getQuantityOfTwoStorageByProductId(this.itemList.get(i).getTypeItem())));
            }
        }
        return remainItemList;
    }
    
    
    
    //
    public short makeOrder(ZPUserInfo user){
        
        for (int i = 0; i < this.itemList.size(); i++){
            if (user.getAsset().getFoodStorage().takeItem(this.itemList.get(i).getTypeItem(), this.itemList.get(i).getQuantity())){
                continue;
            } else if (user.getAsset().getWarehouse().takeItem(this.itemList.get(i).getTypeItem(), this.itemList.get(i).getQuantity())){
                continue;
            } else {
                
                return ErrorLog.ERROR_STORAGE_NOT_REDUCE.getValue();
            }
        }
        
        //
        user.getAsset().getCar().updateDelivery(this.getOrderPrice(), this.getOrderExp());
        //        
        this.waittingTime = 0;
        this.createOrder(user, user.getLevel());
        
        return ErrorLog.SUCCESS.getValue();
    }
    
    
    public short makeOrderByRuby(ZPUserInfo user){
        List<StorageItem> remainItemList = this.checkRemainItem(user);
        for (int i = 0; i < remainItemList.size(); i++){
            if (user.getAsset().getFoodStorage().takeItem(remainItemList.get(i).getTypeItem(), remainItemList.get(i).getQuantity())){
                continue;
            } else if (user.getAsset().getWarehouse().takeItem(remainItemList.get(i).getTypeItem(), remainItemList.get(i).getQuantity())){
                continue;
            } else {
                
                return ErrorLog.ERROR_STORAGE_NOT_REDUCE.getValue();
            }
        }
        
        //
        user.getAsset().getCar().updateDelivery(this.getOrderPrice(), this.getOrderExp());
        //        
        this.waittingTime = 0;
        this.createOrder(user, user.getLevel());
        
        return ErrorLog.SUCCESS.getValue();
    }
    
    
    
    public short cancelOrder(){
        this.itemList = new ArrayList<>();
        this.orderPrice = 0;
        this.orderExp = 0;
        
        this.waittingTime = new Date().getTime();
        
        return ErrorLog.SUCCESS.getValue();
    }
    
    public short boostWait(ZPUserInfo user){
        
        if (user.reduceRuby(3)){
            //        
            this.waittingTime = 0;
            this.createOrder(user, user.getLevel());
            
            return ErrorLog.SUCCESS.getValue();
        }
        
        return ErrorLog.ERROR_RUBY_NOT_REDUCE.getValue();
    }

}
